typed/validate.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	switch {
+	case a.Scalar != nil:
+		return v.doScalar(*a.Scalar, v.value)
+	case a.Struct != nil:
+		return v.doStruct(*a.Struct, v.value)
+	case a.List != nil:
+		return v.doList(*a.List, v.value)
+	case a.Map != nil:
+		return v.doMap(*a.Map, v.value)
+	case a.Untyped != nil:
+		// Untyped sections allow anything.
+		return nil
+	}
+
+	return []ValidationError{v.error("invalid atom")}
+}
+
+func (v validation) doScalar(t schema.Scalar, value value.Value) []ValidationError {
+	switch t {
+	case schema.Numeric:
+		if value.Float == nil && value.Int == nil {
+			// TODO: should the schema separate int and float?
+			return []ValidationError{v.error("expected numeric (int or float)")}
+		}
+	case schema.String:
+		if value.String == nil {
+			return []ValidationError{v.error("expected string")}
+		}
+	case schema.Boolean:
+		if value.Boolean == nil {
+			return []ValidationError{v.error("expected boolean")}
+		}
+	}
+	return nil
+}
+
+func (v validation) doStruct(t schema.Struct, value value.Value) (errs []ValidationError) {
+	switch {
+	case value.Null:
+		// Null is a valid struct.
+		return nil
+	case value.Map != nil:
+		// OK
+	default:
+		return []ValidationError{v.error("expected struct")}
+	}
+
+	allowedNames := map[string]struct{}{}
+	m := *value.Map
+	for i := range t.Fields {
+		f := &t.Fields[i]
+		allowedNames[f.Name] = struct{}{}
+		child, ok := m.Get(f.Name)
+		if !ok {
+			// All fields are optional
+			continue
+		}
+		v2 := v
+		name := f.Name
+		v2.path = append(v.path, fieldpath.PathElement{FieldName: &name})
+		v2.value = child.Value
+		v2.typeRef = f.Type
+		errs = append(errs, v2.validate()...)
+	}
+
+	// All fields may be optional, but unknown fields are not allowed.
+	for _, f := range m.Items {
+		if _, allowed := allowedNames[f.Name]; !allowed {
+			errs = append(errs, v.error("field %v is not mentioned in the schema", f.Name))
+		}
+	}
+
+	// Check unions
+
+	return errs
+}
+
+func listItemToPathElement(list schema.List, index int, child value.Value) (fieldpath.PathElement, error) {
+	pe := fieldpath.PathElement{}
+	if list.ElementRelationship == schema.Associative {
+		if len(list.Keys) > 0 {
+			if child.Map != nil {
+				for _, fieldName := range list.Keys {
+					var fieldValue value.Value
+					field, ok := child.Map.Get(fieldName)
+					if ok {
+						fieldValue = field.Value
+					} else {
+						fieldValue = value.Value{Null: true}
+					}
+					pe.Key = append(pe.Key, value.Field{
+						Name:  fieldName,
+						Value: fieldValue,
+					})
+				}
+				return pe, nil
+			}
+			if child.Null {
+				// Count every key as null?
+				for _, fieldName := range list.Keys {
+					pe.Key = append(pe.Key, value.Field{
+						Name:  fieldName,
+						Value: value.Value{Null: true},
+					})
+				}
+				return pe, nil
+			}
+			return pe, errors.New("associative list with keys has an element that's a non-map type")
+		}
+
+		switch {
```

Comment:
Not completely clear that we are in the "set" mode here, maybe add a comment (or separate into individual functions)

Reply:
will add